internal/controller/controllers/controller_event_wrapper.go

```diff
@@ -0,0 +1,47 @@
+package controllers
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-openapi/strfmt"
+	"github.com/jinzhu/gorm"
+	"github.com/openshift/assisted-service/internal/common"
+	"github.com/openshift/assisted-service/internal/events"
+	"github.com/sirupsen/logrus"
+)
+
+type controllerEventsWrapper struct {
+	events           *events.Events
+	crdEventsHandler CRDEventsHandler
+	db               *gorm.DB
+	log              logrus.FieldLogger
+}
+
+func NewControllerEventsWrapper(crdEventsHandler CRDEventsHandler, events *events.Events, db *gorm.DB, log logrus.FieldLogger) *controllerEventsWrapper {
+	return &controllerEventsWrapper{crdEventsHandler: crdEventsHandler,
+		events: events, db: db, log: log}
+}
+
+func (c *controllerEventsWrapper) AddEvent(ctx context.Context, clusterID strfmt.UUID, hostID *strfmt.UUID, severity string, msg string, eventTime time.Time) {
```

Comment:
I feel like AI events shouldn't used as a trigger for reconciliation.

AI events were meant to notify users about things that happened and be displayed in the UI. I don't think we should use them for control-flow and no component should depend on those events as if they were an interface.

If something happens that requires reconciliation, shouldn't the code just call `NotifyClusterDeploymentUpdates` or `NotifyAgentUpdates` directly? These implicit calls as a result of events feel hacky/surprising.

Reply:
@omertuc the idea of this code is to reconcile on events . It will allow us to stop reconcile every 10 seconds and do it only when needed

Reply:
Event is triggered when something is changed in node/cluster this is the place to trigger reconciliation. I don't think there is any point go through all the code that has event triggering and add  NotifyClusterDeploymentUpdates per each one.

Reply:
This is the way for us to see changes that happen not on user interaction, hosts are added to the cluster, install progress etc.

Reply:
I understand that this solves your problem and I understand how it solves it, I'm not saying it doesn't.

My point is that I don't think events should be used for any purpose other than logs / UI display - this was their purpose all this time, re-purposing them like this feels wrong.

Suddenly they can have real side effects on the behavior of the system - and those real effects are hidden behind an interface polymorphism - it's very surprising.

Reply:
Events are used to show something significant that happened in the cluster, we are using them for debug and collecting metrics as well.

Reply:
@omertuc what is the alternative that you suggest?

Reply:
I already stated my alternative in my original message - I personally think we shouldn't rely on events as a trigger, we should explicitly trigger reconciliation where it's needed. It might look more "cluttered" and it will require more work, but I personally prefer this explicitness.

This is just how I view things, for your consideration - I agree that your solution is the simplest, I'm just afraid it might be confusing/surprising to some people.

But, if you decide to go with your solution - I'm mostly afraid that if someone one day decides to remove an event for whatever reason, then it could potentially delay / completely disable reconciliation (not sure about the exact details here and whether this is realistic, so correct me if I'm wrong).

Due to that, I propose that maybe we should still keep the periodic reconciliation that happens regardless of events, albeit one with a much longer period than the previous 10 second period. 

This minimizes the impact of events from "required for reconciliation" to "makes reconciliation happen earlier than it would've anyway, at best", which is far less likely to cause trouble if some of them get removed some day.

Reply:
I understand your concern, at first we were thinking on using the DB whenever host/cluster is updated we we want to notify the host but because of how gorm is implemented and how we use the DB with the monitor and the periodic requests from the agent it will happen very often.

Another alternative that we considered was to implement a notifier as you suggest but after looking at the case that it is going to be used we saw that it is exactly as the events. we could duplicate the code or wrote it, but it's common in go to use wrappers.

In addition we understand that events become important part of the project and it is used for not only as a nice thing to show the users. 

Regarding breaking things, currently subsystem will break if this feature will not work. so i'm less concerned about it and in worst case if someone add another event then nothing will change and if someone remove events then it will take more time for the users to see the right state but eventually the flow will not break.

Reply:
> if someone remove events then it will take more time for the users to see the right state but eventually the flow will not break

How is that guaranteed? Do reconciliations still happen periodically?

Reply:
Should we discuss this on the architecture call this week?